### DISCUSSION TITLE
Graphs with varying size node features

### DISCUSSION THREAD

**Discussion by tdstr** [pyg-team/pytorch_geometric#7231] [Q&A]
Hi all, I'm trying to work on a delay prediction transport problem, where different services are different nodes in the graph. I'm trying to predict the delay at future stops for a service, given its timetable, and its observed arrival and departure times up to the current timestep.

In a simple case this node data might be:

    Service 1:
        Common features: [Operator, num_stops, service_capacity, etc....]
        Schedule:
            Stops:                  A   B   C   D   E   F
            Expected departure:     13  17  25  27  38  44
            Actual departure:       13  18  28  30  39  ?

    Service 2:
        Common features: [Operator, num_stops, service_capacity, etc....]
        Schedule:
            Stops:                  C   D   E   G    F   H   I   A   B
            Expected departure:     16  29  39  46   60  64  72  81  86
            Actual departure:       16  28  39  43   ?   ?   ?   ?   ?

Where the goal is to predict the times marked by ?. Services are linked if they share any stops within a certain period, and have a higher weight if they share more stops, or if the timetabled departure for the shared stop is closer together.

I'm interested in a graph where services are the nodes, as it should be agnostic of the stop geometry, and hopefully allow for better propagation of delays between different services.

I've looked at works with STGCNS as well as feature propagation for node feature completion but am struggling to make any headway given the varying size of node schedule features.

Will it be necessary to use a heterogenous graph for this or is there any elegant solutions that i'm missing? Similarly, I suspect that encoding the shared stops and the expected time between arrivals in links would yield better performance than just a weight, but again this would result in non uniform features on the links, given different services share different numbers of stops.

Any help or relevant works to read over would be greatly appreciated! Thanks!

**Comment by rusty1s** [comment]
I don't think this necessarily requires a heterogeneous graph. I am not an expert on this topic, but wouldn't it make more sense to see each "Stop" as a node, such that you regress the delay per node (which also removes the need to deal with varying sizes of features).

**Reply by tdstr** [reply]
There has been some work to this effect whereby the structure of the network is encoded that way, and node delays are attributed, which has been successful - The idea of using services as nodes was to hopefully better capture how delays propagate between services, and improve delay prediction for individual services rather than the network sate as a whole. I think i might have a solution where each service is broken down to a cluster of nodes for each stop on the service (retaining homogeneity between nodes), and then the shared stops between different service node clusters are linked. I think that a GAT should be able to account for the different types of links between nodes here as well (same services vs same stop). I'll update this thread if I'm having any success. Thanks so much for your help, your responses on other threads have been invaluable!
